File: objc/objcsearch/objcsearch/FileUtil.m
//
//  FileUtil.m
//  objcsearch
//
//  Created by Cary Clark on 1/14/18.
//  Copyright © 2018 Cary Clark. All rights reserved.
//

#import "FileUtil.h"

@implementation FileUtil

+ (NSSet<NSString*>*) dotDirs {
    return [NSSet setWithObjects:@".", @"..", @"./", @"../", nil];
}

+ (NSString*) expandPath:(NSString*)filePath {
    return [filePath stringByExpandingTildeInPath];
}

+ (NSString*) getExtension:(NSString*)fileName {
    NSURL *fileURL = [NSURL fileURLWithPath:fileName];
    NSString *ext = [[fileURL pathExtension] uppercaseString];
    if ([ext isEqualToString:@"Z"]) {
        return ext;
    }
    return [ext lowercaseString];
}

+ (BOOL) hasExtension:(NSString *)fileName ext:(NSString *)ext {
    return [[self getExtension:fileName] isEqualToString:[ext lowercaseString]];
}

+ (NSFileManager *) getFileManager {
    return [NSFileManager defaultManager];
}

+ (NSArray<NSString*>*) contentsForPath:(NSString*)filePath error:(NSError**)error {
    return [[self getFileManager] contentsOfDirectoryAtPath:filePath error:error];
}

+ (NSDirectoryEnumerationOptions) optionsForSettings:(SearchSettings*)settings {
    NSDirectoryEnumerationOptions options = NSDirectoryEnumerationSkipsPackageDescendants;
    if (settings.excludeHidden) {
        options |= NSDirectoryEnumerationSkipsHiddenFiles;
    }
    if (!settings.recursive) {
        options |= NSDirectoryEnumerationSkipsSubdirectoryDescendants;
    }
    return options;
}

+ (NSDirectoryEnumerator*) enumeratorForPath:(NSString*)filePath settings:(SearchSettings*)settings{
    NSDirectoryEnumerationOptions options = [self optionsForSettings:settings];
    return [[self getFileManager]
            enumeratorAtURL:[NSURL fileURLWithPath:filePath]
            includingPropertiesForKeys:@[NSURLIsDirectoryKey, NSURLIsRegularFileKey]
            options:options
            errorHandler:nil];
}

+ (BOOL) exists:(NSString*)filePath {
    return [[self getFileManager] fileExistsAtPath:filePath];
}

+ (BOOL) isDirectory:(NSString*)filePath {
    if ([[self dotDirs] containsObject:filePath]) {
        return true;
    }
    BOOL isDir;
    if ([[self getFileManager] fileExistsAtPath:filePath isDirectory:&isDir]) {
        return isDir;
    }
    if ([[self getFileManager] fileExistsAtPath:[filePath stringByExpandingTildeInPath] isDirectory:&isDir]) {
        return isDir;
    }
    return false;
}

+ (BOOL) isDotDir:(NSString*)filePath {
    return [[self dotDirs] containsObject:filePath];
}

+ (BOOL) isHidden:(NSString*)filePath {
    NSArray<NSString*> *elems = [filePath componentsSeparatedByString:@"/"];
    for (NSString *elem in elems) {
        if ([self isHiddenFile:elem]) {
            return true;
        }
    }
    return false;
}

+ (BOOL) isHiddenFile:(NSString*)fileName {
    return [fileName hasPrefix:@"."] && ![self isDotDir:fileName];
}

+ (BOOL) isReadableFile:(NSString*)filePath {
    return [[self getFileManager] isReadableFileAtPath:filePath];
}

+ (NSString*) joinPath:(NSString*)path childPath:(NSString*)childPath {
    NSMutableString *joined = [NSMutableString stringWithString:path];
    if (![joined hasSuffix:@"/"]) {
        [joined appendString:@"/"];
    }
    [joined appendString:childPath];
    return [NSString stringWithString:joined];
}

@end

File: swift/swiftsearch/Sources/swiftsearch/FileUtil.swift
//
//  FileUtil.swift
//  swiftsearch
//
//  Created by Cary Clark on 5/18/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

import Foundation

public enum FileUtil {
    fileprivate static let dotDirs = Set<String>([".", "..", "./", "../"])
    fileprivate static let separator = "/"

    public static func contractPath(_ filePath: String) -> String {
        // NOTE: I get this error for the following line when trying to build on linux:
        //       value of type 'NSString' has no member 'abbreviatingWithTildeInPath'
        // (filePath as NSString).abbreviatingWithTildeInPath
        if filePath.hasPrefix(NSHomeDirectory()) {
            return filePath.replacingOccurrences(of: NSHomeDirectory(), with: "~")
        }
        return filePath
    }

    public static func relativePath(_ filePath: String, forPath: String) -> String {
        if forPath == "." || forPath.hasPrefix("./") {
            let fullForPath = URL(fileURLWithPath: ".").path
            return filePath.replacingOccurrences(of: fullForPath, with: ".")
        }
        if forPath == ".." || forPath.hasPrefix("../") {
            let fullForPath = URL(fileURLWithPath: "..").path
            return filePath.replacingOccurrences(of: fullForPath, with: "..")
        }
        return filePath
    }

    // this formats filePath according to forPath, which means that it will become relative
    // if forPath is, or the the HOME prefix will be replaced with tilde if forPath.hasPrefix("~")
    public static func formatPath(_ filePath: String, forPath: String) -> String {
        if forPath.hasPrefix("~") {
            return contractPath(filePath)
        }
        return relativePath(filePath, forPath: forPath)
    }

    public static func formatPath(_ filePath: String, forPaths: [String]) -> String {
        for p in forPaths {
            let formatted: String = formatPath(filePath, forPath: p)
            if formatted.count < filePath.count {
                return formatted
            }
        }
        return filePath
    }

    public static func expandPath(_ filePath: String) -> String {
        (filePath as NSString).expandingTildeInPath
    }

    public static func getExtension(_ fileName: String) -> String {
        let ext = NSURL(fileURLWithPath: fileName).pathExtension?.uppercased()
        if ext == "Z" {
            return ext!
        }
        return ext!.lowercased()
    }

    public static func hasExtension(_ fileName: String, ext: String) -> Bool {
        getExtension(fileName) == ext.lowercased()
    }

    fileprivate static func getFileManager() -> FileManager {
        FileManager.default
    }

    fileprivate static func getOptions(forSettings settings: SearchSettings) -> FileManager.DirectoryEnumerationOptions {
        var options: FileManager.DirectoryEnumerationOptions = [.skipsPackageDescendants]
        if settings.excludeHidden {
            options.insert(.skipsHiddenFiles)
        }
        if !settings.recursive {
            options.insert(.skipsSubdirectoryDescendants)
        }
        return options
    }

    // gets files only directly under given path --> set settings.recursive to false and call `enumerator` below instead
    public static func directoryContents(forPath filePath: String, settings: SearchSettings) -> [String] {
        do {
            let options = getOptions(forSettings: settings)
            let fileUrls = try getFileManager().contentsOfDirectory(at: URL(fileURLWithPath: expandPath(filePath)), includingPropertiesForKeys: [.isRegularFileKey], options: options)
            return fileUrls.map(\.path)
        } catch {
            return []
        }
    }

    // gets files recursively under given path
    public static func enumerator(forPath filePath: String, settings: SearchSettings) -> FileManager.DirectoryEnumerator? {
        let options = getOptions(forSettings: settings)
        return getFileManager().enumerator(at: URL(fileURLWithPath: expandPath(filePath)),
                                           includingPropertiesForKeys: [.isDirectoryKey, .isRegularFileKey],
                                           options: options)
    }

    public static func exists(_ filePath: String) -> Bool {
        getFileManager().fileExists(atPath: expandPath(filePath))
    }

    public static func isDirectory(_ filePath: String) -> Bool {
        var isDir: ObjCBool = false
        if getFileManager().fileExists(atPath: expandPath(filePath), isDirectory: &isDir) {
            return isDir.boolValue
        }
        return false
    }

    public static func isReadableFile(_ filePath: String) -> Bool {
        getFileManager().isReadableFile(atPath: expandPath(filePath))
    }

    public static func isDotDir(_ filePath: String) -> Bool {
        dotDirs.contains(filePath)
    }

    public static func isHidden(_ filePath: String) -> Bool {
        let pathElems = filePath.split { $0 == "/" }.map { String($0) }
        return pathElems.filter { self.isHiddenFile($0) }.count > 0
    }

    public static func isHiddenFile(_ fileName: String) -> Bool {
        fileName.hasPrefix(".") && !isDotDir(fileName)
    }

    public static func splitPath(_ filePath: String) -> (String, String) {
        let pathElems = filePath.split { $0 == "/" }.map { String($0) }
        if pathElems.count > 1 {
            let path = pathElems.prefix(pathElems.count - 1).joined(separator: separator)
            return (path, pathElems[pathElems.count - 1])
        } else {
            return ("", filePath)
        }
    }

    public static func joinPath(_ path: String, childPath: String) -> String {
        if path.hasSuffix(separator) {
            return "\(path)\(childPath)"
        } else {
            return "\(path)/\(childPath)"
        }
    }
}
